drivers/timers/capture.c

```diff
@@ -0,0 +1,377 @@
+
+/* This structure describes the state of the upper half driver */
+
+struct cap_upperhalf_s
+{
+  uint8_t                    crefs;    /* The number of times the device has been opened */
+  sem_t                      exclsem;  /* Supports mutual exclusion */
+  FAR struct cap_lowerhalf_s *lower;   /* lower-half state */
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int     cap_open(FAR struct file *filep);
+static int     cap_close(FAR struct file *filep);
+static ssize_t cap_read(FAR struct file *filep, FAR char *buffer,
+                       size_t buflen);
+static ssize_t cap_write(FAR struct file *filep, FAR const char *buffer,
+                        size_t buflen);
+static int cap_ioctl(FAR struct file *filep, int cmd, unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct file_operations g_capops =
+{
+  cap_open,  /* open */
+  cap_close, /* close */
+  cap_read,  /* read */
+  cap_write, /* write */
+  NULL,      /* seek */
+  cap_ioctl, /* ioctl */
+  NULL,      /* poll */
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+  NULL       /* unlink */
+#endif
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: cap_open
+ *
+ * Description:
+ *   This function is called whenever the PWM Capture device is opened.
+ *
+ ****************************************************************************/
+
+static int cap_open(FAR struct file *filep)
+{
+  FAR struct inode           *inode = filep->f_inode;
+  FAR struct cap_upperhalf_s *upper = inode->i_private;
+  uint8_t                     tmp;
+  int                         ret;
+
+  /* Get exclusive access to the device structures */
+
+  ret = nxsem_wait(&upper->exclsem);
+  if (ret < 0)
+    {
+      goto errout;
+    }
+
+  /* Increment the count of references to the device.  If this is the first
+   * time that the driver has been opened for this device, then initialize
+   * the device.
+   */
+
+  tmp = upper->crefs + 1;
+  if (tmp == 0)
+    {
+      /* More than 255 opens; uint8_t overflows to zero */
+
+      ret = -EMFILE;
+      goto errout_with_sem;
+    }
+
+  /* Check if this is the first time that the driver has been opened. */
+
+  if (tmp == 1)
+    {
+      FAR struct cap_lowerhalf_s *lower = upper->lower;
+
+      /* Yes.. perform one time hardware initialization. */
+
+      DEBUGASSERT(lower->ops->start != NULL);
+
+      ret = lower->ops->start(lower);
```

Comment:
should we let user control start/stop the capture through ioctl? so the user don't need open/close the same device multiple times.

Reply:
Thanks for your agile eyes! Maybe we can start capture through open,because the capture driver is different from pwm driver, the pwm driver should be started through ioctl when system is ready, It's dangerous if pwm output at device open stage. There's no such things for capture driver, if no pulse at the beginning, the capture will not be triggered, and if there's pulse at the beginning,the results will be overwritten. In fact, I implemented capture driver by reference to ./drivers/sensors/hall3ph.c

Reply:
Look reasonable.